src/index.ts

```diff
@@ -1568,44 +1565,81 @@ export class BigQuery extends common.Service {
    */
   query(
     query: string | Query,
-    optionsOrCallback?:
-      | QueryOptions
-      | SimpleQueryRowsCallback
-      | QueryRowsCallback,
-    cb?: SimpleQueryRowsCallback | QueryRowsCallback
-  ): void | Promise<SimpleQueryRowsResponse> | Promise<QueryRowsResponse> {
-    let options =
+    optionsOrCallback?: QueryOptions | QueryRowsCallback,
+    cb?: QueryRowsCallback
+  ): void | Promise<QueryRowsResponse> {
+    const options =
       typeof optionsOrCallback === 'object' ? optionsOrCallback : {};
     const callback =
       typeof optionsOrCallback === 'function' ? optionsOrCallback : cb;
-    this.createQueryJob(query, (err, job, resp) => {
-      if (err) {
-        (callback as SimpleQueryRowsCallback)(err, null, resp);
-        return;
-      }
-      if (typeof query === 'object' && query.dryRun) {
-        (callback as SimpleQueryRowsCallback)(null, [], resp);
-        return;
-      }
-      // The Job is important for the `queryAsStream_` method, so a new query
-      // isn't created each time results are polled for.
-      options = extend({job}, options);
-      job!.getQueryResults(options, callback as QueryRowsCallback);
-    });
+
+    query = typeof query === 'string' ? {query} : query;
+
+    if (query.table) {
+      query.table.getRows(options, callback!);
+      return;
+    }
+
+    this.runQuery_(query, options).then(
+      ([rows, nextQuery, resp]) => callback!(null, rows, nextQuery, resp),
+      callback!
+    );
   }
 
   /**
-   * This method will be called by `createQueryStream()`. It is required to
-   * properly set the `autoPaginate` option value.
+   * It is faster to use the `listTableData` rpc instead of `getQueryResults`,
+   * so when we run a query, we perform the following steps
+   *
+   *   1. create the query job
+   *   2. poll `getQueryResults` with `maxResults` set to 0 to check for job
+   *      completion
+   *   3. if we have a positive row count, call `listTableData` on the
+   *      destination table created by query job
    *
    * @private
    */
-  queryAsStream_(query: Query, callback?: SimpleQueryRowsCallback) {
-    if (query.job) {
-      query.job.getQueryResults(query, callback as QueryRowsCallback);
-      return;
+  async runQuery_(
+    query: Query,
+    options: QueryOptions
+  ): Promise<QueryRowsResponse> {
+    const [job, resp] = await this.createQueryJob(query);
+
+    if (query.dryRun) {
+      return [[], null, resp];
+    }
+
+    const queryOptions = Object.assign({}, options, {
+      autoPaginate: false,
+      maxResults: 0,
+    });
+
+    let rows, nextQuery, results;
+
+    do {
+      [rows, nextQuery, results] = await job.getQueryResults(queryOptions);
```

Comment:
@tswast does the Python client use any kind of backoff in between `getQueryResults` calls?

Reply:
Yes, in two ways.

1. The default retry exponential backoff for each individual call:

   https://github.com/googleapis/google-cloud-python/blob/master/bigquery/google/cloud/bigquery/job.py#L2793-L2829
  
   Note: we set a timeout of 10 seconds per call. The call hangs until timeout or the job completes.
2. `result()` checks the value of `done()` until it's true or the call times out.

   https://github.com/googleapis/google-cloud-python/blob/83f0262fc52bcf63b57e3902d83656c70a7cbcc2///api_core/google/api_core/future/polling.py#L87-L105

Reply:
@callmehiphop do we currently, or need to, do any of that here?

Reply:
The default retry is baked in from the common libraries, but we could probably do a better job of implementing step 2.